libs/datamodel/connectors/sql-datamodel-connector/src/postgres_datamodel_connector.rs

```diff
@@ -42,30 +118,137 @@ impl Connector for PostgresDatamodelConnector {
+                    PostgresType::Bit(arg)
+                }
+            }
+            VAR_BIT_TYPE_NAME => {
+                if let Some(arg) = *args.first() {
+                    PostgresType::VarBit(arg)
+                }
             }
+            UUID_TYPE_NAME => PostgresType::UUID,
+            JSON_TYPE_NAME => PostgresType::JSON,
+            JSON_B_TYPE_NAME => PostgresType::JSONB,
             _ => unreachable!("This code is unreachable as the core must guarantee to just call with known names."),
         };
 
-        Some(NativeTypeInstance::new(constructor.name.as_str(), args, &native_type))
+        let native_type_constructor = self.constructors.iter().find(|c| c.name == name)?;
+
+        if native_type_constructor._number_of_args != length {
+            return Err(ConnectorError::new_argument_count_mismatch_error(
+                name,
+                native_type_constructor._number_of_args,
+                length,
+            ));
+        }
+
+        // check for compatability with scalar type
+        let compatable_prisma_scalar_type = native_type_constructor.prisma_type;
+        if compatable_prisma_scalar_type != scalar_type {
+            return Err(ConnectorError::new_incompatible_native_type_error(
+                "Postgres",
+                scalar_type,
+                compatable_prisma_scalar_type,
+            ));
+        }
+
+        match constructor {
+            Some(constructor) => Ok(NativeTypeInstance::new(constructor.name.as_str(), args, &native_type)),
+            _ => panic!(""),
+        }
     }
 
-    fn introspect_native_type(&self, native_type: Box<dyn NativeType>) -> Option<NativeTypeInstance> {
+    fn introspect_native_type(&self, native_type: Box<dyn NativeType>) -> Result<NativeTypeInstance, ConnectorError> {
         let native_type: PostgresType = serde_json::from_value(native_type.to_json()).unwrap();
         let (constructor_name, args) = match native_type {
+            PostgresType::SmallInt => (SMALL_INT_TYPE_NAME, vec![]),
+            PostgresType::Integer => (INTEGER_TYPE_NAME, vec![]),
             PostgresType::BigInt => (BIG_INT_TYPE_NAME, vec![]),
+            PostgresType::Real => (REAL_TYPE_NAME, vec![]),
+            PostgresType::DoublePrecision => (DOUBLE_PRECISION_TYPE_NAME, vec![]),
+            PostgresType::SmallSerial => (SMALL_SERIAL_TYPE_NAME, vec![]),
+            PostgresType::Serial => (SMALL_SERIAL_TYPE_NAME, vec![]),
+            PostgresType::BigSerial => (BIG_SERIAL_TYPE_NAME, vec![]),
             PostgresType::VarChar(x) => (VARCHAR_TYPE_NAME, vec![x]),
-            _ => todo!("This match must be exhaustive"),
+            PostgresType::Char(x) => (CHAR_TYPE_NAME, vec![x]),
+            PostgresType::Text => (TEXT_TYPE_NAME, vec![]),
+            PostgresType::Timestamp(x) => (TIMESTAMP_TYPE_NAME, vec![x]),
+            PostgresType::TimestampWithTimeZone(x) => (TIMESTAMP_WITH_TIMEZONE_TYPE_NAME, vec![x]),
```

Comment:
Given this is a 2-element array, not sure if that adds more complexity here and other places.

```suggestion
            PostgresType::TimestampWithTimeZone(x) => (TIMESTAMP_WITH_TIMEZONE_TYPE_NAME, vec![x]),
            PostgresType::TimestampWithTimeRangeZone(x) => (TIMESTAMP_WITH_TIMEZONE_RANGE_TYPE_NAME, vec![x]),
```